src/main/java/io/jenkins/plugins/postgresql/PostgreSQLFingerprintStorage.java

```diff
@@ -0,0 +1,334 @@
+ * THE SOFTWARE.
+ */
+package io.jenkins.plugins.postgresql;
+
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import com.thoughtworks.xstream.converters.basic.DateConverter;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.model.TaskListener;
+import jenkins.fingerprints.FingerprintStorage;
+import hudson.model.Fingerprint;
+import hudson.Util;
+
+import java.io.IOException;
+import java.io.ByteArrayInputStream;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.json.JSONArray;
+import org.jenkinsci.main.modules.instance_identity.InstanceIdentity;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+/**
+ * Pluggable external fingerprint storage for fingerprints in PostgreSQL.
+ */
+@Extension
+public class PostgreSQLFingerprintStorage extends FingerprintStorage {
+
+    private final String instanceId;
+    private static final DateConverter DATE_CONVERTER = new DateConverter();
+    private static final Logger LOGGER = Logger.getLogger(PostgreSQLFingerprintStorage.class.getName());
+
+    public static PostgreSQLFingerprintStorage get() {
+        return ExtensionList.lookupSingleton(PostgreSQLFingerprintStorage.class);
+    }
+
+    @DataBoundConstructor
+    public PostgreSQLFingerprintStorage() throws IOException {
+        instanceId = Util.getDigestOf(new ByteArrayInputStream(InstanceIdentity.get().getPublic().getEncoded()));
+    }
+
+    public Connection getConnection() throws SQLException {
+        return PostgreSQLConnection.getConnection(getHost(), getPort(), getDatabaseName(), getCredentialsId(), getSsl(),
+                getConnectionTimeout(), getSocketTimeout());
+    }
+
+    /**
+     * Saves the given fingerprint.
+     */
+    public synchronized void save(Fingerprint fingerprint) {
+        delete(fingerprint.getHashString());
+        try (Connection connection = getConnection()) {
+            connection.setAutoCommit(false);
+
+            PreparedStatement preparedStatement = connection.prepareStatement(Queries.getQuery("insert_fingerprint"));
+            preparedStatement.setString(1, fingerprint.getHashString());
+            preparedStatement.setString(2, instanceId);
+            preparedStatement.setString(3, DATE_CONVERTER.toString(fingerprint.getTimestamp()));
+            preparedStatement.setString(4, fingerprint.getFileName());
+            preparedStatement.execute();
+            preparedStatement.close();
+
+            Hashtable<String, Fingerprint.RangeSet> usages = fingerprint.getUsages();
```

Comment:
Hmmmm....yikes, on Hashtable. I wonder how hard it would be change the Fingerprint class to a modern Hash

Reply:
Modern hash as in something like `ConcurrentHashMap`? That does seem like it would be better, but it would be a change in Jenkins core. TBH I think it might not be a trivial change because backward compatibility would be a key issue here.